### DISCUSSION TITLE
has_table and views in 1.4-branch

### DISCUSSION THREAD

**Discussion by ffyring** [sqlalchemy/sqlalchemy#8678] [Ideas]
Somewhere along the 1.4 updates has_table stopped checking for views for MSSQL (dialects/mssql/base.py).

I checked in the master branch, and there it was fixed with commit de61582933b800272cc818d36de9e0b7f86aa4fe. It is not fixed in the 1.4 branch however.

I need to use the anaconda-supplied version (last version 1.4.42, seems to be a monthly release schedule) and I guess next conda version will be 1.4.43 a month or so from now. 

@CaselIT  Is it possible to port or commit same thing to the 1.4-branch also?

**Comment by CaselIT** [comment]
Hi,

I don't think 1.4 ever reflected views for MSSQL. What was the the version that reflected views in the 1.4 series?

I don't think it's appropriate to change the behaviour of has_table in 1.4 at the moment unless a regression removed a functionality that was previously there

**Reply by ffyring** [reply]
Hi,
has_table stopped working with views in 1.4.0, and worked in 1.3.x. In 1.4.0 you explicitely check for a table
 ```
s = sql.select(tables.c.table_name).where(
                sql.and_(
                    tables.c.table_type == "BASE TABLE",
                    tables.c.table_name == tablename,
                )
            )
```
In main branch it can be a table or a view
```

  s = sql.select(tables.c.table_name).where(
                sql.and_(
                    sql.or_(
                        tables.c.table_type == "BASE TABLE",
                        tables.c.table_type == "VIEW",
                    ),
                    tables.c.table_name == tablename,
                )
            )
```
The explicit check from version 1.4 causes a regresssion, and adding the "or"-clause from main branch would fix that.

**Reply by CaselIT** [reply]
Interesting. In 1.3 it was really an has_object since there was no filter.

@zzzeek how do you want to handle this?

**Reply by ffyring** [reply]
@zzzeek  I want to add that important projects depend on this: currently pandas.to_sql(view) is broken and treating views as tables is an important use-case in mssql

**Reply by zzzeek** [reply]
has_table() works for views and can confirm it stopped doing that for mssql specifically in 1.4 so this is a regression, #8700

**Reply by zzzeek** [reply]
does not work for Oracle however and that's not a "regression".   A sudden change here late in the 1.4 series could potentially break Oracle code that relies on this retunring False.  it may be unlikely but 1.4 is very late.

**Reply by CaselIT** [reply]
on v2 it will work also for oracle, since there are now tests that verify that has_table will also return views

**Reply by zzzeek** [reply]
take a look at the review, you'll see
